### PR TITLE
[RPC] Add coldstaking address support in importprivkey and importaddress

### PR DESCRIPTION
**Pull request by random-zebra**
Still it remains to add proper support in `dumpwallet` and `importwallet`

### CODE REVIEW

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -197,8 +201,12 @@ UniValue importaddress(const UniValue& params, bool fHelp)
             throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this address or script");
 
         // add to address book or update label
-        if (address.IsValid())
-            pwalletMain->SetAddressBook(address.Get(), strLabel, AddressBook::AddressBookPurpose::RECEIVE);
+        if (address.IsValid()) {
+            std::string strPurpose = AddressBook::AddressBookPurpose::RECEIVE;
+            if (address.IsStakingAddress())
+                strPurpose = AddressBook::AddressBookPurpose::COLD_STAKING;
```
Comment (furszy) [inline src/wallet/rpcdump.cpp]: Minor nit, inline this strPurpose initialization.

File: src/wallet/rpcdump.cpp [hunk 2]
```diff
@@ -102,30 +103,28 @@ UniValue importprivkey(const UniValue& params, bool fHelp)
...
+    if (!key.IsValid())
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Private key outside allowed range");
 
     CPubKey pubkey = key.GetPubKey();
     assert(key.VerifyPubKey(pubkey));
     CKeyID vchAddress = pubkey.GetID();
     {
         pwalletMain->MarkDirty();
-        pwalletMain->SetAddressBook(vchAddress, strLabel, AddressBook::AddressBookPurpose::RECEIVE);
+        pwalletMain->SetAddressBook(vchAddress, strLabel, (
```
Comment (furszy) [inline src/wallet/rpcdump.cpp]: Move the ```LOCK2(cs_main, pwalletMain->cs_wallet);``` that is at the beginning of the method inside this block to not lock anything if the data is invalid.

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (CaveSpectre11) [reply src/wallet/rpcdump.cpp]: or (minor nit) use same code style as DecodeDumpString()
```
        if (address.IsValid()) {
                pwalletMain->SetAddressBook(address.Get(), strLabel, address.IsStakingAddress() ?
                        AddressBook::AddressBookPurpose::COLD_STAKING :
                        AddressBook::AddressBookPurpose::RECEIVE));
            }

**Comment by random-zebra** [comment]
Uniformed code styling for `strPurpose` in the two calls and moved the lock. Squashed.

**Comment by furszy** [comment]
Would be nice to add this scenario to the `wallet_import_rescan.py` or well, create a new test to verify the import + rescan for staking addresses.

The PR is good, if you want to tackle the functional test in another PR let me know and will ack it.

**Comment by random-zebra** [comment]
Good idea. 
Though `wallet_import_rescan.py` is one of the currently broken tests. It should be fixed but in a separate PR. For now will add just a single test for these changes.

**Comment by random-zebra** [comment]
Tests added.

File: src/wallet/rpcdump.cpp [hunk 3]
```diff
@@ -75,16 +75,17 @@ std::string DecodeDumpString(const std::string& str)
 
 UniValue importprivkey(const UniValue& params, bool fHelp)
 {
-    if (fHelp || params.size() < 1 || params.size() > 3)
+    if (fHelp || params.size() < 1 || params.size() > 4)
         throw std::runtime_error(
             "importprivkey \"pivxprivkey\" ( \"label\" rescan )\n"
```
Comment (Fuzzbawls) [inline src/wallet/rpcdump.cpp]: nit: add new `fStakingAddress` optional argument to this line

Comment (random-zebra) [reply src/wallet/rpcdump.cpp]: yep. done

**Pull request merged**